Issue1442 load coupling

AntoineGautier opened the pull request:
@mwetter Could you please review this PR to URBANopt staging branch? (The CI testing is limited to the package `Buildings.Applications` for now.)
It includes:

- base models: `Buildings.Applications.DHC.Loads.Examples.BaseClasses` and `Buildings.Applications.DHC.Networks.BaseClasses` 
- and validation and example models: `Buildings.Applications.DHC.Loads.Examples` and `Buildings.Applications.DHC.Loads.Validation`.

The full DHC system models based on your previous work on RN is in [issue1769_examples5GDHC](https://github.com/lbl-srg/modelica-buildings/tree/issue1769_examples5GDHC) but the development in that branch is not finalized yet, see #1769.

Eventually this branch has a dependency to `issue1129_energyPlus_zone` that will first be merged into `issue1437_district_heating_cooling` as discussed.

AntoineGautier commented:
@mwetter I realized there are models from the wet coil development branch in that branch. I will close this PR and push a new clean branch. It that OK from your side?

AntoineGautier commented:
@mwetter As discussed I:

- renamed the geoJSON export (and cleaned it as some files were not used in that branch),
- merged `issue1129_energyPlus_zone` into `issue1437_district_heating_cooling`.

You can start the review.

(I will open issues on `geojson-modelica-translator` for the rounding, air medium and building code name.)

mwetter commented on Buildings/Applications/DHC/Loads/Validation/BaseClasses/FanCoil2PipeCooling.mo:
```diff
@@ -0,0 +1,186 @@
+  if have_fluPor then
+  end if;
+  if not have_QReq_flow then
+  end if;
```
Empty if-then statements

mwetter commented on Buildings/Applications/DHC/Loads/Validation/BaseClasses/FanCoil2PipeHeating.mo:
```diff
@@ -0,0 +1,199 @@
+air temperature is assessed with
+<a href=\"modelica://Buildings.Applications.DHC.Loads.BaseClasses.SimpleRoom\">
+Buildings.Applications.DHC.Loads.BaseClasses.SimpleRoom</a>.
+</p>
```
Missing revision notes.

mwetter commented on Buildings/Applications/DHC/Loads/Validation/BaseClasses/FanCoil2PipeHeatingValve.mo:
```diff
@@ -0,0 +1,208 @@
+  if have_fluPor then
+  end if;
+  if not have_QReq_flow then
+  end if;
```
Empty if-then

mwetter commented on Buildings/Applications/DHC/Loads/Validation/BenchmarkFlowDistribution1.mo:
```diff
@@ -0,0 +1,156 @@
+      StopTime=2000000,
+      __Dymola_NumberOfIntervals=500,
+      Tolerance=1e-06,
+      __Dymola_Algorithm="Cvode"),
```
Remove __Dymola_NumberOfIntervals and __Dymola_Algorithm in all files.

mwetter commented on Buildings/Applications/DHC/Loads/Validation/BenchmarkFlowDistribution1.mo:
```diff
@@ -0,0 +1,156 @@
+    annotation (Placement(transformation(extent={{-100,60},{-80,80}})));
+  Buildings.Controls.OBC.UnitConversions.From_degC minTSet_K
+    "Minimum temperature setpoint (K)"
+    annotation (Placement(transformation(extent={{-60,60},{-40,80}})));
```
Please remove From_degC and use instead
```
minTSet(
  k(
  final unit="K",
  "displayUnit="degC") =293.15)
```

AntoineGautier commented on Buildings/Applications/DHC/Controls/MixingValveControl.mo:
```diff
@@ -0,0 +1,224 @@
+    connect(fixMod.y, swi.u2)
+      annotation (Line(points={{11,0},{68,0}},
+                     color={255,0,255}));
+  end if;
```
You are right: not needed here.

AntoineGautier commented on Buildings/Applications/DHC/Loads/BaseClasses/FlowDistribution.mo:
```diff
@@ -0,0 +1,866 @@
+        annotation (Line(points={{-100,0},{-16,0}}, color={0,127,255}));
+    end if;
+
+  end if;
```
You are right, I could clean up several equation sections. A few conditional statements are still needed for this reason: statements involving conditionally removed components are removed at translation time by Modelica specification. However, obsolete statements corresponding to the default model structure need to be programmatically removed. I included that comment in the equation section when needed.

AntoineGautier commented:
@mwetter Thanks a lot for the detailed review. I addressed all your comments (here above and on paper). There are two remaining points to agree on.

- I wanted to add a validation case for the `PartialBuilding` configuration where `PHea` and `PCoo` are used, i.e., for decentralized heating and cooling systems like split systems. However, implementing a split model is not as straightforward as I expected (cf. our discussion by email). I still would like to merge the model as is because I do not see where the model could fail considering the simplicity of that option.
```
  parameter Boolean have_eleHea = true
    "Set to true if the building has decentralized electric heating equipment"
    annotation(Evaluate=true);
...
  Modelica.Blocks.Interfaces.RealOutput PHea(
    final quantity="Power", final unit="W") if have_eleHea
    "Power drawn by decentralized heating equipment"
```

- I agree that the logic for modeling a change-over terminal unit is not intuitive (`have_watCoo = true`, `have_chaOve = true`, and `have_watHea = false`). That is why I tried to document it thoroughly. The reason for that situation is the use of two pairs of fluid ports for heating and chilled water, instead of a fluid port array. One of these two pairs needs to be removed for a change-over system. Keeping the chilled water label makes more sense to me as, fundamentally, a change-over coil is a chilled water coil (with condensate draining) configured to also operate with heating water.

If we agree on that then this is ready to merge when all test pass.

mwetter commented:
@AntoineGautier The first point (merging without a test for a split system) is fine, although we could approximate a sensible only split system with one of the `HeatPumps.Carnot_*`.

Let's discuss the second point on today's call. What I was thinking is an enumeration such as the one below (types are just examples)
```
  type HeatingCooling = enumeration(
      ChangeOverHydronicHydronicHeating "Change over system with DX coil and hydronic heating",
      ChangeOverReversibleHeatPump "Changer over system with reversible heat pump",
      HydronicHeatingOnly "Only hydronic heating",
      HydronicCoolingOnly "Only hydronic cooling")
```
Then, we would assign the `have_*` parameters based on this enumeration using `final`. Other systems could be added in the future as needed.

AntoineGautier commented:
@mwetter This is ready to merge when all tests pass.

This closes #1442 and URBANopt [#35](https://github.com/urbanopt/urbanopt-modelica/issues/35).

mwetter commented:
@AntoineGautier : I don't see why this one does not run the CI tests at https://github.com/lbl-srg/modelica-buildings/pull/1730 as another one from this morning runs: https://travis-ci.com/github/lbl-srg/modelica-buildings/builds/155572411

If you tested it locally, please feel free to merge.
